src/main/java/com/redhat/pantheon/model/api/SlingResource.java

```diff
@@ -0,0 +1,274 @@
+     */
+    protected void initDefaultValues() {
+        getMembers(Field.class)
+                // discard fields which already have a value set
+                .filter(field -> !field.isSet())
+                // only consider fields which have a default value
+                .filter(field -> field.defaultValue.isPresent())
+                .forEach(field -> field.set(field.defaultValue.get()));
+    }
+
+    /**
+     * Returns all SlingModel members (assigned fields which implement the ResourceMember interdace
+     * @return
+     */
+    private Stream<ResourceMember> allMembers() {
+        return stream(this.getClass().getDeclaredFields())
+                // only fields which implement ResourceMember
+                .filter(reflectedField -> ResourceMember.class.isAssignableFrom(reflectedField.getType()))
+                // convert to the field values
+                .map(field -> {
+                    try {
+                        return (ResourceMember) field.get(SlingResource.this);
+                    } catch (IllegalAccessException e) {
+                        throw new RuntimeException(e);
+                    }
+                })
+                // only initialized fields
+                .filter(field -> field != null);
+    }
+
+    /**
+     * Returns all SlingModel members of a specific implementation.
+     */
+    private <A extends ResourceMember> Stream<A> getMembers(@Nonnull Class<A> memberClass) {
+        return (Stream<A>) allMembers()
+                .filter(resourceMember -> memberClass.isAssignableFrom(resourceMember.getClass()));
+    }
+
+    /**
+     * Returns a map with all the fields (deep fields included) for this model. The keys for the map
+     * are the jcr field names.
+     *
+     * @param excluding A list of JCR field names to exclude from the returned map.
+     * @return
+     */
+    public Map<String, Object> toMap(String ... excluding) {
+        // get the fields first
+        Map<String, Object> returnMap = getMembers(Field.class)
+                .filter(field ->
+                        stream(excluding).noneMatch(arrayField -> arrayField.equals(field.getName())))
+                .collect(Collectors.toMap(
+                        field -> field.getName(),
+                        field -> field.get()
+                ));
+
+        // add the deep fields
+        returnMap.putAll(
+                getMembers(DeepField.class)
+                        .filter(deepField ->
+                            stream(excluding).noneMatch(arrayField -> arrayField.equals(deepField.getPath())))
+                        // get rid of null values
+                        .filter(deepField -> deepField.get() != null)
+                        .collect(Collectors.toMap(
+                                deepField -> deepField.getPath(),
+                                deepField -> deepField.get()
+                        ))
+        );
```

Comment:
Because of the difference in API. `DeepField` has `getPath` and `Field` has `getName` which means I can't use the same chained calls. Java doesn't have pattern matching which would have solved this, and I guess an alternative would be to reflectively ask about the specific type, cast it and then operate on it, which is kind of nasty, but slightly more efficient (we are talking negligible levels of efficiency here, both would be O(n)).

Reply:
Ah, didn't notice the getName vs getPath at for the first go-round. Fair enough.